python_ta/debug/accumulation_table.py

```diff
@@ -0,0 +1,133 @@
+        self._num_iterations += 1
+
+    def _create_iteration_dict(self) -> dict:
+        """Use the values of each iteration and return a dictionary that maps each accumulator
+        and loop variable to its respective value during each iteration
+        """
+        table_dict = {
+            "iteration": list(range(self._num_iterations + 1)),
+            "loop variable (" + self._loop_var_name + ")": self._loop_var_val,
+            **self._loop_accumulation_values,
+        }
+        return table_dict
+
+    def _tabulate_data(self) -> None:
+        """Print the values of the accumulator and loop variables into a table"""
+        print(
+            tabulate.tabulate(
+                self._create_iteration_dict(), headers="keys", colalign=("left", "left", "left")
+            )
+        )
+
+    def _no_of_whitespaces(self, with_line: str) -> int:
+        """Return the number of indents that are in the line of code containing the with statement"""
+        blank_chars = 0
+        for char in with_line:
+            if char == " ":
+                blank_chars += 1
+            else:
+                break
+
+        return blank_chars
+
+    def _loop_string(self, full_string: list[str], no_whitespaces: int) -> str:
+        """Convert the list of lines that are strings from the beginning of the for loop
+        to the end of the for loop"""
+        endpoint = len(full_string)
+        for i in range(len(full_string)):
+            if full_string[i] != "" and not (
+                full_string[i][no_whitespaces] == " " or full_string[i][no_whitespaces] == "\t"
+            ):
+                endpoint = i
+                break
+
+        return "\n".join(full_string[:endpoint])
+
+    def _add_keys(self, func_node: types.NodeNG, func_frame: types.FrameType) -> None:
+        """Add the names and values of the all the accumulators for the zeroth iteration"""
+        self._loop_var_name = func_node.target.name
+        for name in self._accumulation_names:
+            if name in func_frame.f_locals:
+                self._loop_accumulation_values[name] = [func_frame.f_locals[name]]
+            else:
+                raise NameError
+
+    def _trace_loop(self, frame: types.FrameType, event: str, arg: Any) -> None:
+        if event == "line" and frame.f_lineno == self._curr_lineno:
+            local_vars = frame.f_locals
+            curr_vals = []
+            for accumulator in self._accumulation_names:
+                if accumulator in local_vars:
+                    curr_vals.append(local_vars[accumulator])
+            if self._loop_var_name in local_vars:
+                self._curr_values = [curr_vals, local_vars[self._loop_var_name]]
+                self._add_iteration(curr_vals, local_vars[self._loop_var_name])
+
+    def __enter__(self) -> AccumulationTable:
+        self._loop_var_val.append("N/A")
+        func_frame = inspect.getouterframes(inspect.currentframe())[1].frame
```

Comment:
The code from here to line 122 should be moved into a (well-documented) helper function